### PR TITLE
refactor: Scheduler/Calendar page/component

### PR DESCRIPTION
**Pull request by aomi**
# Description
This PR takes refactors some of the calendar code that was speedily coded in order to release on time. There are several bits of code where we can likely move to the backend but this PR just refactors it so we can pull pieces out easier in the future.

## Checklist

- [x] The code follows all style guidelines.
- [x] The code passes all required tests.
- [x] The code is documented.
- [ ] The code includes tests.
- [x] I have self-reviewed my changes and have done QA.

## General Comments
Apart from possible speed improvements due to caching some parsing operations, it should function the same as before.

### CODE REVIEW

File: src/pages/scheduler/components/Event.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { Flex, Heading, HStack, VStack } from '@chakra-ui/react';
+import { EventProps } from 'react-big-calendar';
+
+export const CustomEvent = ({ title, event }: EventProps) => {
```
Comment (keithradford) [inline src/pages/scheduler/components/Event.tsx:4]: the name `CustomEvent` seems a bit redundant and I'd say something like `CalendarEvent` is more clear

File: src/pages/scheduler/components/Event.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  return (
+    <Flex height="100%" direction="column">
+      <HStack w="100%" bg="#EDF2F7" justifyContent="space-between" p="0.2em">
```
Comment (keithradford) [inline src/pages/scheduler/components/Event.tsx:7]: could probably move away from hard coding hexes in, even if Chakra doesn't offer the _exact_ same shade. Also, maybe dark mode integration 😉  ?

File: src/pages/scheduler/components/SchedulerCalendar.tsx [hunk 3]
```diff
@@ -1,233 +1,176 @@
...
+  const lowerBound = parse(event.term, 'yyyyMM', new Date());
+
+  const startEndDates = event.meetingTime.dateRange.split('-').map((d) => d.replace(',', ''));
+  const startEndTimes = event.meetingTime.time.split('-').map((d) => d.trim());
+
+  // TODO: find better means of handling timezones
+  const courseStartDate = new Date(startEndDates[0] + ' 00:00:00 GMT');
+  const courseEndDate = new Date(startEndDates[1] + ' 00:00:00 GMT');
+
+  // TODO: why is it adding +1 to the month?
```
Comment (keithradford) [inline src/pages/scheduler/components/SchedulerCalendar.tsx:72]: one month type might be indexed at 0 and the other 1?

File: src/pages/scheduler/components/Toolbar.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import { ChevronLeftIcon, ChevronRightIcon } from '@chakra-ui/icons';
+import { Button, Flex, Heading, HStack, IconButton, Text } from '@chakra-ui/react';
+import { ToolbarProps } from 'react-big-calendar';
+
+export const CustomToolBar = (onSelectedDateChange: (date: Date) => void) => ({ label, date }: ToolbarProps) => {
+  const handleClick = (offset?: number) => () => {
```
Comment (keithradford) [inline src/pages/scheduler/components/Toolbar.tsx:6]: probably aware but a bunch of lint errors here

File: src/pages/scheduler/components/Toolbar.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import { ChevronLeftIcon, ChevronRightIcon } from '@chakra-ui/icons';
+import { Button, Flex, Heading, HStack, IconButton, Text } from '@chakra-ui/react';
+import { ToolbarProps } from 'react-big-calendar';
+
+export const CustomToolBar = (onSelectedDateChange: (date: Date) => void) => ({ label, date }: ToolbarProps) => {
```
Comment (keithradford) [inline src/pages/scheduler/components/Toolbar.tsx:5]: again, maybe something like `CalendarToolbar` is more clear

File: src/pages/scheduler/containers/SchedulerContainer.tsx [hunk 6]
```diff
@@ -6,49 +6,26 @@ import { useParams } from 'react-router';
...
+  const calendarEvents = useMemo(
+    () => denormalizeCourseEvents(coursesResult.status === 'loaded' ? coursesResult.data : []),
+    [coursesResult]
+  );
 
   return (
     <Flex grow={1} height="100%" overflow="hidden">
       <Box w="100%" height="100%" px="3" py="2">
-        <SchedulerCalendar calendarEvents={calendarEvents} />
+        {coursesResult.status}
```
Comment (keithradford) [inline src/pages/scheduler/containers/SchedulerContainer.tsx:27]: Don't know if this is supposed to be here 👀

File: src/pages/scheduler/hooks/useCalendarEvents.tsx [hunk 7]
```diff
@@ -20,16 +18,17 @@ type CalendarEventsResult =
...
-  const [result, setResult] = useState<CalendarEventsResult>({ status: 'loading' });
+  const [result, setResult] = useState<GetCourseSectionsResult>({ status: 'loading' });
 
   // avoid unnessary filtering operations by memoization
   const termCourses = useMemo(() => courses.filter((c) => c.term === term), [term, courses]);
 
   useEffect(() => {
     (async () => {
       try {
+        // setResult({ status: 'loading' });
```
Comment (keithradford) [inline src/pages/scheduler/hooks/useCalendarEvents.tsx:31]: Can probably be removed

File: src/pages/scheduler/hooks/useTransformedCalendarEvents.tsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { CourseCalendarEvent, SavedCourseWithSections } from 'pages/scheduler/shared/types';
+
+// denormalizes the calendar events from the saved courses
```
Comment (keithradford) [inline src/pages/scheduler/hooks/useTransformedCalendarEvents.tsx:3]: what's denormalize?

Comment (aomi) [reply src/pages/scheduler/hooks/useTransformedCalendarEvents.tsx:3]: > Denormalization is a strategy used on a previously-normalized database to increase performance. In computing, denormalization is the process of trying to improve the read performance of a database, **at the expense of losing some write performance, by adding redundant copies of data or by grouping data**.

File: src/pages/scheduler/components/Event.tsx (hunk 2 above)
Comment (aomi) [reply src/pages/scheduler/components/Event.tsx]: you love dark mode don't you :wink: done.

File: src/pages/scheduler/components/Toolbar.tsx [hunk 9]
```diff
@@ -2,39 +2,41 @@ import { ChevronLeftIcon, ChevronRightIcon } from '@chakra-ui/icons';
...
+        <Heading size="md">Scheduler</Heading>
+        <Text fontSize="xl">{label}</Text>
+        <HStack pb="0.2em">
+          <Button size="sm" colorScheme="gray" onClick={handleClick()}>
+            Today
+          </Button>
+          <IconButton
+            aria-label="Previous Week"
+            bg="gray"
+            icon={<ChevronLeftIcon color="white" />}
```
Comment (keithradford) [inline src/pages/scheduler/components/Toolbar.tsx:28]:
```suggestion
            colorScheme="gray"
            icon={<ChevronLeftIcon />}
```

**Review by keithradford — APPROVED** [review]
One minor thing, but LGTM

**Pull request merged**